### PR TITLE
gprecoverseg: allow failover to a new host while preserving partial successes

### PR DESCRIPTION
**Pull request by dkrieger-pvtl**
We had introduced [functionality to ignore segments on unreachable hosts](https://github.com/greenplum-db/gpdb/commit/c880732f0d4090fe75ba1aa586317bf65823bc6d) but still recover remaining segments on reachable hosts.  However, this unintentionally broke "gprecoverseg -p newhost" behavior.

This PR only supports the case without "-i config_file.txt" specified; that is "gprecoverseg -p newhost" or "gprecoverseg".  That case will be handled in a separate PR.

Most of this PR is establishing test cases for the CI in Concourse in a new Behave feature "gprecoverseg_newhost", as we require multiple hosts.  As we cannot easily shutdown and reboot a CCP node in Concourse, we instead isolate a machine using blackhole routes.  This makes said host unreachable on the network, simulating a physically shutdown host.

[pipeline](https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-cm-recoverseg_new_host_177765775)

### CODE REVIEW

**Review by kalensk — COMMENTED** [review]
I would consider removing references to gphostcache and isStandardArray from gprecoverseg. Or worse case as long as that on the radar for right away and not slip through the cracks.

**Comment by kalensk** [comment]
I would suggest updating the implementation in `clsRecoverSegment.py` to filter out the unreachable hosts in `source_recovery_hosts` and `target_recovery_hosts` as generally shown below. This removes the need `failoverSegment.unreachable = False` and is more logical to follow and cleaner. The checks in `GpMirrorToBuild()` would remain.


In `getRecoveryActionsFromConfiguration()`:
```

    # Check if source recovery hosts are reachable
    source_recovery_hosts = [seg.getSegmentHostName() for seg in peersForFailedSegments]
    unreachable_hosts = get_unreachable_segment_hosts(source_recovery_hosts, len(source_recovery_hosts))
    if unreachable_hosts:
        logger.warning("Cannot recover. The recovery source host %s is unreachable." % (' '.join(map(str, unreachable_hosts))))
    # NOTE: Noeed to filter out the source_recovery_hosts (ie: peersForFailedSegments), put keep all the reachable ones and proceed

    # Check if target recovery hosts are reachable
    target_recovery_hosts = list(zip(*recoverAddressMap.values())[0])
    unreachable_hosts = get_unreachable_segment_hosts(target_recovery_hosts, len(target_recovery_hosts))
    if unreachable_hosts:
        logger.warning("Cannot recover. The recovery target host %s is unreachable." % (' '.join(map(str, unreachable_hosts))))
    # NOTE: Noeed to filter out the unreachable target_recover_hosts (ie: recoverAddressMap), put keep all the reachable ones and proceed
    ```

**Comment by kalensk** [comment]
Can you clearly state in the PR or commit message why we are not addressing `getRecoveryActionsFromConfigFile`?

**Comment by dkrieger-pvtl** [comment]
> For future reference, can you clearly state in the PR and/or commit message why we are not addressing `getRecoveryActionsFromConfigFile`? This would help people when they need to figure out why in the future.

I updated that commit.

**Comment by dkrieger-pvtl** [comment]
> I would suggest updating the implementation in `clsRecoverSegment.py` to filter out the unreachable hosts in the `source_recovery_hosts` and `target_recovery_hosts` as generally shown below. This removes the need `failoverSegment.unreachable = False` and is more logical to follow and cleaner. The checks in `GpMirrorToBuild()` would also be removed since unreachable hosts would not get there.

In principle, I agree but the current implementation isn't clean enough to do it that simply.  If we are not recovering to a new host, `failoverSegment` is set to `None` here and then a triple (failedSegment, liveSegment, None) is passed down the call stack to recover this triple.  In this path(from gparray), that should mean it is recovered on the same host as `failedSegment`.  But the code to do that is long and far away...so I'd prefer to keep the structure as it is for now.

For the upcoming refactor, the first thing I would like to do is to combine the two paths(FromCongFile and FromConfiguration)  into a set of "recovery triples" that get passed down after filtering out those that are not possible due to a corresponding host being unreachable.

**Comment by dkrieger-pvtl** [comment]
> The checks in `GpMirrorToBuild()` would also be removed since unreachable hosts would not get there.

The added checks(that the source liveSegment and target failoverSegment are both on reachable hosts) are still appropriate in that function.  That function basically asserts facts about the (failedhost, liveSegment, failoverSegment) that we _thought_ we handled.  In some sense, these checks should have been added as soon as we added the functionality to determine unreachable hosts.

File: concourse/tasks/run_behave_on_ccp_cluster.yml [hunk 1]
```diff
@@ -71,6 +71,18 @@ run:
...
+    if (( $SUDO_ACCESS_ON_HOSTS )); then
+        while read LINE; do
+            host=$(echo $LINE | awk '{ print $3 }')
+            ssh -n $SUDO_MDW_USERNAME@$host "
+              sudo usermod -a -G google-sudoers gpadmin
+            "
+        done <cluster_env_files/etc_hostfile
+    fi
+
     # start cluster back up
```
Comment (dkrieger-pvtl) [inline concourse/tasks/run_behave_on_ccp_cluster.yml:86]: expand leading comment on L74 to state the current use(to run `ip route add`).

File: concourse/pipelines/templates/gpdb-tpl.yml [hunk 2]
```diff
@@ -1269,12 +1274,34 @@ jobs:
...
+              outputs:
+                - name: cluster_env_files
+              run:
+                path: sh
+                args:
+                  - -exc
+                  - |
+                    # modify the hostfile_init used by gpinitsystem to exclude the last two hosts, so they are spares
+                    cp cluster_env_files/hostfile_init cluster_env_files/hostfile_init_orig
+                    head -$(( $(cat cluster_env_files/hostfile_init_orig | wc -l) - 2 )) cluster_env_files/hostfile_init_orig > cluster_env_files/hostfile_init
```
Comment (dkrieger-pvtl) [inline concourse/pipelines/templates/gpdb-tpl.yml:1294]: is there a cleaner way to pull all but the last two lines of the file?

File: gpMgmt/test/behave/mgmt_utils/gprecoverseg_newhost.feature [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      And the mirrors replicate and fail over and back correctly
+      And segment hosts <down> are reconnected to the cluster and to the spare segment hosts "<unused>"
+      And the original cluster state is recreated to <down>
+      And database "gptest" exists
+      And the cluster configuration is saved "after_recreation"
+      And the "before" and "after_recreation" cluster configuration matches with the expected for gprecoverseg newhost
+      Examples:
+      | test_case  |  down        | spare       | unused |   when    | gprecoverseg_cmd                  | down_sql                                              |
+      | host_one   |  "sdw1"      | "sdw5,sdw6" | sdw6   | after_one | "gprecoverseg -a -p sdw5"      | "hostname='sdw1' and status='u'"                      |
+      | host_two   |  "sdw1,sdw3" | "sdw5,sdw6" | none   | after_two | "gprecoverseg -a -p sdw5,sdw6" | "(hostname='sdw1' or hostname='sdw3') and status='u'" |
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/gprecoverseg_newhost.feature:32]: "<when>" is hard to understand...be more explicit like `one_host_down` or add a comment

File: gpMgmt/test/behave/mgmt_utils/gprecoverseg_newhost.feature [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      And the cluster configuration is saved "before"
+      And segment hosts <down> are disconnected from the cluster and from the spare segment hosts <spare>
+      And the cluster configuration has no segments where <down_sql>
+      When the user runs <gprecoverseg_cmd>
+      Then gprecoverseg should return a return code of 0
+      And the cluster configuration is saved "<when>"
+      And the "before" and "<when>" cluster configuration matches with the expected for gprecoverseg newhost
+      And the mirrors replicate and fail over and back correctly
+      And segment hosts <down> are reconnected to the cluster and to the spare segment hosts "<unused>"
+      And the original cluster state is recreated to <down>
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/gprecoverseg_newhost.feature:25]: And the original cluster state is recreated after cleaning up down hosts `<down>`

File: gpMgmt/test/behave/mgmt_utils/steps/unreachable_hosts_mgmt_utils.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+#           hosts                            isolate_hosts      spare
+# ['mdw', 'sdw1', 'sdw2', 'sdw3', 'sdw4'], ['sdw1', 'sdw3'], ['sdw5', 'sdw6'] (disconnect)
+# ['mdw', 'sdw2', 'sdw4', 'sdw5', 'sdw6'], ['sdw1', 'sdw3'], [])              (reconnect)
+def add_or_remove_blackhole_route(isolated, spare, disconnect=False):
+    hosts = GpArray.initFromCatalog(dbconn.DbURL()).getHostList()
+    hosts.extend(spare)
+
+    if disconnect:
+        # sdw1 -x- sdw3  and  sdw3 -x- sdw1
+        for isolate in isolated:
```
Comment (dkrieger-pvtl) [inline gpMgmt/test/behave/mgmt_utils/steps/unreachable_hosts_mgmt_utils.py:43]: use `for host in isolated_hosts`

Comment (dkrieger-pvtl) [reply gpMgmt/test/behave/mgmt_utils/steps/unreachable_hosts_mgmt_utils.py:43]: or disconnected and connected

**Review by kalensk — APPROVED** [review]
Given the minor renames looks good.

File: concourse/pipelines/templates/gpdb-tpl.yml (hunk 2 above)
Comment (dkrieger-pvtl) [reply concourse/pipelines/templates/gpdb-tpl.yml:1294]: Done.

File: concourse/tasks/run_behave_on_ccp_cluster.yml (hunk 1 above)
Comment (dkrieger-pvtl) [reply concourse/tasks/run_behave_on_ccp_cluster.yml:86]: Done.

File: gpMgmt/test/behave/mgmt_utils/gprecoverseg_newhost.feature (hunk 3 above)
Comment (dkrieger-pvtl) [reply gpMgmt/test/behave/mgmt_utils/gprecoverseg_newhost.feature:32]: Done...replaced use of "when" with a reworded "test_case".

File: gpMgmt/test/behave/mgmt_utils/steps/unreachable_hosts_mgmt_utils.py (hunk 5 above)
Comment (dkrieger-pvtl) [reply gpMgmt/test/behave/mgmt_utils/steps/unreachable_hosts_mgmt_utils.py:43]: I renamed "isolated" to "disconnected" and generally improved the phrasing.

**Pull request merged**